Emby.Server.Implementations/Session/SessionManager.cs

```diff
@@ -1841,88 +1832,67 @@ public Task SendMessageToUserSessions<T>(List<Guid> userIds, string name, Func<T
-                }
-                catch (Exception ex)
+                foreach (var session in sessions)
                 {
-                    _logger.LogError("Error sending message", ex);
+                    yield return SendMessageToSession(session, name, data, cancellationToken);
                 }
+            }
 
-            }, cancellationToken)).ToArray();
-
-            return Task.WhenAll(tasks);
+            return Task.WhenAll(GetTasks());
         }
 
         public Task SendMessageToUserDeviceSessions<T>(string deviceId, string name, T data, CancellationToken cancellationToken)
         {
             CheckDisposed();
 
-            var sessions = Sessions.Where(i => string.Equals(i.DeviceId, deviceId, StringComparison.OrdinalIgnoreCase)).ToList();
+            var sessions = Sessions.Where(i => string.Equals(i.DeviceId, deviceId, StringComparison.OrdinalIgnoreCase));
 
-            var tasks = sessions.Select(session => Task.Run(async () =>
+            IEnumerable<Task> GetTasks()
             {
-                try
-                {
-                    await SendMessageToSession(session, name, data, cancellationToken).ConfigureAwait(false);
-                }
-                catch (Exception ex)
+                foreach (var session in sessions)
                 {
-                    _logger.LogError("Error sending message", ex);
+                    yield return SendMessageToSession(session, name, data, cancellationToken);
                 }
+            }
 
-            }, cancellationToken)).ToArray();
-
-            return Task.WhenAll(tasks);
+            return Task.WhenAll(GetTasks());
         }
 
         public Task SendMessageToUserDeviceAndAdminSessions<T>(string deviceId, string name, T data, CancellationToken cancellationToken)
         {
             CheckDisposed();
 
             var sessions = Sessions
-                .Where(i => string.Equals(i.DeviceId, deviceId, StringComparison.OrdinalIgnoreCase) || IsAdminSession(i))
-                .ToList();
+                .Where(i => string.Equals(i.DeviceId, deviceId, StringComparison.OrdinalIgnoreCase) || IsAdminSession(i));
 
-            var tasks = sessions.Select(session => Task.Run(async () =>
+            IEnumerable<Task> GetTasks()
             {
-                try
-                {
-                    await SendMessageToSession(session, name, data, cancellationToken).ConfigureAwait(false);
-                }
-                catch (Exception ex)
+                foreach (var session in sessions)
                 {
-                    _logger.LogError("Error sending message", ex);
+                    yield return SendMessageToSession(session, name, data, cancellationToken);
                 }
+            }
 
-            }, cancellationToken)).ToArray();
-
-            return Task.WhenAll(tasks);
+            return Task.WhenAll(GetTasks());
```

Comment:
I see this pattern of

1. getting a sessions list
2. defining a `GetTasks` which calls `SendMessageToSession` for all sessions
3. returning `Task.WhenAll(GetTasks())`

repeated at least three times. Can we extract points 2 and 3 in a separate (probably static) method to reduce copy-pasteness?

Reply:
Done 👍